manta-accounting/src/wallet/signer/mod.rs

```diff
@@ -835,6 +846,217 @@ where
+)]
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = "Asset<C>: Clone, Identifier<C>: Clone"),
+    Debug(bound = "Asset<C>: Debug, Identifier<C>: Debug"),
+    Eq(bound = "Asset<C>: Eq, Identifier<C>: Eq"),
+    Hash(bound = "Asset<C>: Hash, Identifier<C>: Hash"),
+    PartialEq(bound = "Asset<C>: PartialEq, Identifier<C>: PartialEq")
+)]
+pub struct ConsolidationRequest<C>(Vec<IdentifiedAsset<C>>)
+where
+    C: transfer::Configuration + ?Sized;
+
+impl<C> ConsolidationRequest<C>
+where
+    C: transfer::Configuration,
+    IdentifiedAsset<C>: PartialEq,
+{
+    /// Builds a new [`ConsolidationRequest`] from `assets`.
+    #[inline]
+    fn new_unchecked(assets: Vec<IdentifiedAsset<C>>) -> Self {
+        Self(assets)
+    }
+
+    /// Builds a new [`ConsolidationRequest`] from `assets`, ensuring that:
+    /// 1) There are at least two [`IdentifiedAsset`]s in `assets`.
+    /// 2) There are no duplicates in `assets`.
+    /// 3) All assets in `assets` have the same asset id.
+    /// 4) No asset in `assets` has zero value.
+    #[inline]
+    pub fn new(assets: Vec<IdentifiedAsset<C>>) -> Option<Self> {
+        let number_of_assets = assets.len();
+        if number_of_assets < 2 {
+            return None;
+        }
+        let asset_id = &assets
+            .first()
+            .expect("This cannot fail because of the check above.")
+            .asset
+            .id;
+        (!(1..number_of_assets).any(|i| {
+            assets[i..].contains(&assets[i - 1])
+                && assets[i].asset.id.ne(asset_id)
+                && assets[i].asset.value.eq(&Default::default())
+        }))
+        .then_some(Self::new_unchecked(assets))
+    }
+
+    /// Returns the asset id of the assets in `self`.
+    #[inline]
+    pub fn id(&self) -> &C::AssetId {
+        &self
+            .0
+            .first()
+            .expect("Consolidation requests must have at least 2 elements")
+            .asset
+            .id
+    }
+
+    /// Returns the total amount of [`Asset`] held by the assets in `self`.
+    #[inline]
+    pub fn asset(&self) -> Asset<C> {
+        Asset::<C>::new(
+            self.id().clone(),
+            self.0
+                .iter()
+                .map(|identified_asset| identified_asset.asset.value.clone())
+                .sum(),
+        )
+    }
+
+    /// Checks that every asset in `self` is also in `asset_map`.
+    #[inline]
+    pub fn check_consolidation_request<M>(&self, asset_map: &M) -> bool
+    where
+        M: AssetMap<C::AssetId, C::AssetValue, Key = Identifier<C>>,
+    {
+        let id = &self
+            .0
+            .first()
+            .expect("Consolidation requests must have at least 2 elements")
+            .asset
+            .id;
+        let asset_map_assets = asset_map
+            .asset_vector_with_id(id)
+            .into_iter()
+            .map(|(identifier, asset)| IdentifiedAsset::<C>::new(identifier, asset))
+            .collect::<Vec<_>>();
```

Comment:
How big is this Vector ? Maybe we should collect it into a hash set because of the following lookup

Reply:
The vector shouldn't be too big. Also, `HashSet` is "forbidden" for wallet implementations (I got some trouble with the WASM part of the sdk when there were `std` calls).

Reply:
could write our own i guess lol (in my free time maybe)